modules/layouts/LayoutCS.chpl

```diff
@@ -0,0 +1,656 @@
+  proc key(idx: _tuple) { return (idx(2), idx(1));}
+}
+
+const _columnComparator: _ColumnComparator;
+
+/*
+This CS layout provides a Compressed Sparse Row and Compressed Sparse Column
+implementation for Chapel's sparse domains and arrays.
+
+To declare a CS domain, invoke the ``CS`` constructor without arguments
+in a `dmapped` clause. For example:
+
+  .. code-block:: chapel
+
+    use LayoutCS;
+    var D = {1..n, 1..m};  // a default-distributed domain
+    var CS_Domain: sparse subdomain(D) dmapped CS();
+
+To declare a CS array, use a CS domain, for example:
+
+  .. code-block:: chapel
+
+    // assume the above declarations
+    var CS_Array: [CS_Domain] real;
+
+This domain map is a layout, i.e. it maps all indices to the current locale.
+All elements of a CS-distributed array are stored
+on the locale where the array variable is declared.
+*/
+class CS: BaseDist {
+  const row: bool = true;
+
+  proc dsiNewSparseDom(param rank: int, type idxType, dom: domain) {
+    return new CSDom(rank=rank, row=row, idxType=idxType, dist=this, parentDom=dom);
+  }
+
+  proc dsiClone() return new CS(row=this.row);
+
+  proc dsiEqualDMaps(that: CS) param {
+    return true;
+  }
+
+  proc dsiEqualDMaps(that) param {
+    return false;
+  }
+} // CS
+
+
+class CSDom: BaseSparseDomImpl {
+  const row: bool;                  // would like param
+
+  var dist: CS;
+
+  var rowRange: range(idxType);
+  var colRange: range(idxType);
+
+  /* (row|col) dom */
+  const dom: domain(1, idxType);
+
+  /* (row|col) start */
+  var startIdx: [dom] idxType;      // would like index(nnzDom)
+  /* (row|col) idx */
+  var idx: [nnzDom] idxType;        // would like index(parentDom.dim(1))
+
+  /* Initializer */
+  proc CSDom(param rank, type idxType, const row=true,
+             dist: CS, parentDom: domain(rank, idxType)) {
+    if (rank != 2) then
+      compilerError("Only 2D sparse domains are supported by the CS distribution");
+    this.row = row;
+    this.dist = dist;
+    this.parentDom = parentDom;
+    rowRange = parentDom.dim(1);
+    colRange = parentDom.dim(2);
+    dom = if row then {rowRange.low..rowRange.high+1} else {colRange.low..colRange.high+1};
+    nnzDom = {1..nnz};
+    dsiClear();
+  }
+
+  proc dsiMyDist() return dist;
+
+  // TODO: Open question -- How should slicing on CS domains and arrays work?
+  proc dsiGetIndices() { compilerError('Slicing is not supported on CS domains'); }
+
+  proc dsiSetIndices(x) { compilerError('Slicing is not supported on CS domains'); }
+
+  proc dsiAssignDomain(rhs: domain, lhsPrivate:bool) {
+    chpl_assignDomainWithIndsIterSafeForRemoving(this, rhs);
+  }
+
+  proc dsiBuildArray(type eltType)
+    return new CSArr(eltType=eltType, rank=rank, idxType=idxType, dom=this);
+
+  iter dsiIndsIterSafeForRemoving() {
+    var cursor = if this.row then rowRange.high else colRange.high;
+    for i in 1..nnz by -1 {
+      while (startIdx(cursor) > i) {
+        cursor -= 1;
+      }
+      if this.row {
```

Comment:
This line is a great example of why we'd want a param -- without that, all our iterators become "two yield" iterators which we're much worse at optimizing than those with a single yield.  If 'row' is a param, we'll fold this conditional and have a single yield.

If we can't make row a param for some reason, we should at least store the index to be yielded into a variable in a conditional and then yield that variable to get back to a single-yield world.